### PR TITLE
Restructure Single Symmetry into Global and Local Symmetry

### LINKED ISSUE #240: Disambiguate crystal symmetry symbols
Depending on the choice of origin / setting, the H-M and Hall symbols will change. Hall records the changes that were applied, H-M doesn't. Technically, one could produce a list of equivalent symbols, especially for H-M.

There are standard (short) representations for H-M and Hall, which is what most people will use when trying to search for a crystal at large. We should also decide whether rotations and shifts should be applied to show the crystal in its standard setting.

### PR DESCRIPTION
**Pull request by ndaelman-hu**
# Purpose

1. Introduces two polymorphic subsections into `ModelSymmetry` / `Representation` to handle global / local symmetries. The split exists as these properties transform under different conditions. There were predecessors PRs, but due to the introduction of `Representation`, had to be redone.
2. Extend on the pre-existing crystal symmetries and analysis.

Closes #240 and #242

Scope

Included (What is intentionally part of this PR?)

- Schema updates for `ModelMethod`
- Testing for the analysis and sanity checking logic
- Documentation (at the quantity, section, and repo level)
- Backwards compatibility?

Out of scope (What is explicitly not included (to avoid scope creep)?)

- Parser compatibility updates

Status

Select one

 - [ ] Draft / In progress (not ready for full review)
 - [x] Ready for review
 - [ ] Blocked (explain below)

Dependencies / Blockers

List anything that may block progress or review

- [x] None.
- [ ] Related PRs:
- [ ] Open design questions:
- [ ] External dependencies (people, decisions, data, releases):

Testing / Validation

- [ ] Not applicable: schema-only change; no runtime behavior to exercise yet.
- [x] Unit tests
- [x] Integration tests
- [ ] Manual testing (brief description):
- [ ] Not applicable (explain why):

Linked Issues / Context

- Closes #240 and #242 
- Follow-up issues (if known): #104, #219

Reviewer Notes

Please provide feedback on:

- [ ] Software architectural change
- [ ] Need for backwards compatibility
- [ ] Crystal symmetry analysis (if you feel comfortable with it)
- [ ] Test edge cases

### CODE REVIEW

File: src/nomad_simulations/schema_packages/model_system.py [hunk 1]
```diff
@@ -323,25 +306,191 @@ class AlternativeRepresentation(Representation):
...
+
+            # Check each populated array for any quantity with shape ['*']
+            for field_name, quantity in self.m_def.all_quantities.items():
+                if quantity.shape and quantity.shape[0] == '*':
+                    field_value = getattr(self, field_name, None)
+                    if field_value is not None and len(field_value) != n_particles:
+                        logger.warning(
+                            f'{self.__class__.__name__}.{field_name} length ({len(field_value)}) '
+                            f'does not match n_particles ({n_particles})'
+                        )
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/model_system.py]: move out + test

File: src/nomad_simulations/schema_packages/model_system.py [hunk 2]
```diff
@@ -424,13 +582,123 @@ class Symmetry(ArchiveSection):
...
+        distinct from generic representation transformations.
+        """,
+    )
+
     atomic_cell_ref = Quantity(
         type=Representation,
         description="""
         Reference to the Representation section that the symmetry refers to.
         """,
     )
```
Comment (ndaelman-hu) [inline src/nomad_simulations/schema_packages/model_system.py:433]: redundant?

**Comment by ndaelman-hu** [comment]
I verified that the parsers in `nomad-parser-plugins-simulation` are fully compatible with this symmetry refactoring. All 51 parser tests pass without any code changes required.

The parsers do not directly instantiate or manipulate symmetry objects - they populate atomic structure data and rely on the schema's normalization functions to perform symmetry analysis automatically. The backward compatibility measures in this PR (single vs. repeating subsection) work seamlessly with the existing parser implementations.

**Review by ndaelman-hu — COMMENTED** [review]
**Re: Pearson notation parsing for 2D/1D lattices:**

This is a known limitation that I will address when I add full support for 2D/1D lattice systems in a future PR. For now, the schema only handles 3D crystal systems with single-character Pearson family codes (a, m, o, t, r, h, c). I'll implement multi-character parsing logic (mp, op, tp, hp, ap) when extending to lower-dimensional systems.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 3]
```diff
@@ -424,13 +656,167 @@ class Symmetry(ArchiveSection):
...
+            'S': 'S - face centred',
+            'R': 'R - rhombohedral',
+            'c': 'c - centered 2D',
+            'p': 'p - primitive 2D/1D',
+        }
+
+        # Extract family and centering from Pearson notation
+        family_code = pearson[0].lower()
+        centering_code = pearson[1] if len(pearson) > 1 else ''
+
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: This is a known limitation that I will address when I add full support for 2D/1D lattice systems in a future PR. For now, the schema only handles 3D crystal systems with single-character Pearson family codes (a, m, o, t, r, h, c). I'll implement multi-character parsing logic (mp, op, tp, hp, ap) when extending to lower-dimensional systems.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 4]
```diff
@@ -527,19 +953,95 @@ def resolve_bulk_symmetry(
...
+                    else [conventional_wyckoff]
+                )
+                conventional_letter_counts = {
+                    letter: conv_wyckoff_list.count(letter)
+                    for letter in set(conv_wyckoff_list)
+                }
+                # Map each original position to its multiplicity
+                site_mults = [
+                    conventional_letter_counts.get(letter, 1)
+                    for letter in original_wyckoff
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: Fixed in commit 28a7d58. The calculation now uses `equivalent_atoms` grouping instead of counting Wyckoff letters. I extracted the logic into a reusable `_compute_site_multiplicities()` static method that's used in both `resolve_bulk_symmetry()` and `resolve_analyzed_cell()`. Added comprehensive unit tests with 6 parametrized cases including the ZnS wurtzite scenario you mentioned.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 5]
```diff
@@ -491,6 +877,46 @@ def resolve_analyzed_cell(
...
+                        )
+                        conventional_letter_counts = {
+                            letter: conv_wyckoff_list.count(letter)
+                            for letter in set(conv_wyckoff_list)
+                        }
+                        site_mults = [
+                            conventional_letter_counts.get(letter, 1)
+                            for letter in wyckoff
+                        ]
+                        cell_section.local_symmetry.site_multiplicities = site_mults
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: Fixed in commit 28a7d58 using the same `_compute_site_multiplicities()` method. Both locations now call this shared method instead of duplicating the logic.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 6]
```diff
@@ -323,25 +306,265 @@ class AlternativeRepresentation(Representation):
...
+            # Check each populated array for any quantity with shape ['*']
+            for field_name, quantity in self.m_def.all_quantities.items():
+                if quantity.shape and quantity.shape[0] == '*':
+                    field_value = getattr(self, field_name, None)
+                    if field_value is not None and len(field_value) != n_particles:
+                        logger.warning(
+                            f'{self.__class__.__name__}.{field_name} length ({len(field_value)}) '
+                            f'does not match n_particles ({n_particles})'
+                        )
+
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py:396]: Fixed in commit 28a7d58. Added `test_local_symmetry_array_length_validation()` to test the warning path when array lengths don't match the parent representation's particle count.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 7]
```diff
@@ -323,25 +306,265 @@ class AlternativeRepresentation(Representation):
...
+        Returns:
+            list[str] | None: List of Wyckoff site annotations in format "letter+multiplicity",
+            or None if either wyckoff_letters or site_multiplicities is not set.
+
+        Examples:
+            - ['a1', 'b2', 'b2', 'c4', 'c4', 'c4', 'c4'] indicates:
+              • 1 atom at Wyckoff position 'a' (special position, multiplicity 1)
+              • 2 symmetrically equivalent atoms at Wyckoff position 'b' (multiplicity 2)
+              • 4 symmetrically equivalent atoms at Wyckoff position 'c' (general position, multiplicity 4)
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: Fixed in commit 28a7d58. Removed the incorrect "general position" label from the example. Position 'c' with multiplicity 4 is now correctly described without the general/special distinction.

File: tests/test_model_system.py [hunk 8]
```diff
@@ -1398,3 +1397,137 @@ def test_root_representative_flag_is_preserved_on_normalize(self):
...
+        pbc=True,
+    )
+
+    sys = ModelSystem.from_ase_atoms(ase_atoms, logger=logger)
+    sys.type = 'bulk'
+    symmetry = Symmetry()
+
+    # Directly call resolve_bulk_symmetry to test the implementation
+    primitive_cell, conventional_cell = symmetry.resolve_bulk_symmetry(sys, logger)
+
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply tests/test_model_system.py:1516]: Fixed in commit 28a7d58. Changed to `_, _ = symmetry.resolve_bulk_symmetry(sys, logger)` to signal we're intentionally discarding the returned cells.

File: tests/test_model_system.py [hunk 9]
```diff
@@ -1398,3 +1397,137 @@ def test_root_representative_flag_is_preserved_on_normalize(self):
...
+        cell=[a, a, a],
+        pbc=True,
+    )
+
+    sys = ModelSystem.from_ase_atoms(ase_atoms, logger=logger)
+    sys.type = 'bulk'
+    symmetry = Symmetry()
+
+    # Directly call resolve_bulk_symmetry to test the implementation
+    primitive_cell, conventional_cell = symmetry.resolve_bulk_symmetry(sys, logger)
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply tests/test_model_system.py]: Fixed in commit 28a7d58. Changed to `_, _ = symmetry.resolve_bulk_symmetry(sys, logger)` to signal we're intentionally discarding the returned cells.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 10]
```diff
@@ -424,13 +656,167 @@ class Symmetry(ArchiveSection):
...
+        # Extract the letter codes from the MEnum values
+        # MEnum format is "X - description", we need just the first character(s)
+        family_code = self.lattice_type.split(' - ')[0] if self.lattice_type else ''
+        centering_code = (
+            self.lattice_centering.split(' - ')[0] if self.lattice_centering else ''
+        )
+
+        return (
+            f'{family_code}{centering_code}' if family_code and centering_code else None
+        )
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: Not addressing this in the current PR. The code already handles 3D Pearson notation correctly (tested indirectly via `test_symmetry_analysis_fields`). Full support for 2D/1D lattices will be added in a future PR when we extend the schema for lower-dimensional systems. Testing dead code for unreachable 2D/1D cases has negative ROI.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 11]
```diff
@@ -1467,9 +1969,9 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
         #         self.fractional_coordinates = self.compute_fractional_coordinates()
 
         # Create and normalize Symmetry section if applicable
-        if self.type == 'bulk' and self.symmetry is not None:
-            sec_symmetry = self.m_create(Symmetry)
-            sec_symmetry.normalize(archive, logger)
+        if self.type == 'bulk' and self.symmetry is None:
+            self.symmetry = GlobalCrystalSymmetry()
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py:2006]: This is the intended behavior. The condition creates  for bulk systems when it doesn't already exist, which ensures proper symmetry analysis during normalization.

At best, we could add more settings for entrypoints to control this behavior if needed. @JFRudzinski - thoughts on whether we need entrypoint-level control here?

File: src/nomad_simulations/schema_packages/model_system.py [hunk 12]
```diff
@@ -323,25 +306,265 @@ class AlternativeRepresentation(Representation):
...
+            for field_name, quantity in self.m_def.all_quantities.items():
+                if quantity.shape and quantity.shape[0] == '*':
+                    field_value = getattr(self, field_name, None)
+                    if field_value is not None and len(field_value) != n_particles:
+                        logger.warning(
+                            f'{self.__class__.__name__}.{field_name} length ({len(field_value)}) '
+                            f'does not match n_particles ({n_particles})'
+                        )
+
+
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: Fixed in commit c1f153f. Added fallback chain: fractional_coordinates → positions → n_particles. This makes validation robust for ModelSystem instances that may only have positions populated.

File: src/nomad_simulations/schema_packages/model_system.py [hunk 13]
```diff
@@ -424,13 +656,167 @@ class Symmetry(ArchiveSection):
...
+        # MEnum format is "X - description", we need just the first character(s)
+        family_code = self.lattice_type.split(' - ')[0] if self.lattice_type else ''
+        centering_code = (
+            self.lattice_centering.split(' - ')[0] if self.lattice_centering else ''
+        )
+
+        return (
+            f'{family_code}{centering_code}' if family_code and centering_code else None
+        )
+
```
[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: The property already handles multi-character family codes correctly. The implementation uses `split(' - ')[0]` which extracts the full prefix before the separator, whether it's single-character (3D: "c", "t", etc.) or multi-character (2D/1D: "mp", "op", etc.).

Updated the docstring to clarify this explicitly with examples for both cases.

[the remark this answers is not in the document]
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py:849]: Fixed in commit aa220c1. The parsing logic now correctly handles both single-character (3D) and multi-character (2D/1D) family codes by checking for two-character codes first, then falling back to single characters.

The `bravais_lattice` property already handled both cases correctly via `split(' - ')[0]`.

Added comprehensive test coverage with 28 parametrized test cases covering parsing and round-trip reconstruction for both 3D and 2D/1D Pearson notation.

File: src/nomad_simulations/schema_packages/model_system.py (hunk 1 above)
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py]: Fixed in commit 9a5c302. Extracted validation logic into helper methods:
- `_get_particle_count_from_parent()`: determines particle count from parent representation
- `_validate_array_lengths()`: validates all variable-length arrays match particle count

Added comprehensive test coverage with 7 parametrized test cases.

File: src/nomad_simulations/schema_packages/model_system.py (hunk 2 above)
Comment (ndaelman-hu) [reply src/nomad_simulations/schema_packages/model_system.py:433]: Fixed in commit 9a5c302. Removed the deprecated `atomic_cell_ref` field completely.

Investigation confirmed it was entirely unused:
- No assignments anywhere in schema/parsers/NOMAD-FAIR
- No reads anywhere in codebase
- Documentation explicitly stated it "provides no functionality"

Also updated documentation files that incorrectly referenced it as functional.

**Review by JFRudzinski — APPROVED** [review]
Looks good to me :raised_hands:
